From d7a4caf4cc59fbcac4dab8d4ac2c87193d1ffe0f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C3=86var=20Arnfj=C3=B6r=C3=B0=20Bjarmason?= <avarab@gmail.com>
Date: Mon, 21 Jun 2010 18:05:12 +0000
Subject: [PATCH] checkout: Better error message on invalid branch name
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Change the error message for `git checkout -b "I LOVE YOU"' and other
invalid branch names to say what branch names are invalid, rather than
just saying we don't like it.

Before:

    $ git checkout -b "I LOVE YOU"
    fatal: git checkout: we do not like 'I LOVE YOU' as a branch name.

After:

    $ git checkout -b "I LOVE YOU"
    fatal: git checkout: The name "I LOVE YOU" is not a valid Git branch name.
    A branch name can not:
        - Have a path component that begins with "."
        - Have a double dot ".."
        - Have an ASCII control character, "~", "^", ":" or SP, anywhere
        - End with a "/"
        - End with ".lock"
        - Contain a "\" (backslash

Signed-off-by: Ævar Arnfjörð Bjarmason <avarab@gmail.com>
---

Git didn't like my branch name, but I didn't like its error
message. Here's a fix for it.

 builtin/checkout.c |   10 +++++++++-
 1 files changed, 9 insertions(+), 1 deletions(-)

diff --git a/builtin/checkout.c b/builtin/checkout.c
index 1994be9..fec8335 100644
--- a/builtin/checkout.c
+++ b/builtin/checkout.c
@@ -856,7 +856,15 @@ no_reference:
 	if (opts.new_branch) {
 		struct strbuf buf = STRBUF_INIT;
 		if (strbuf_check_branch_ref(&buf, opts.new_branch))
-			die("git checkout: we do not like '%s' as a branch name.",
+			die("git checkout: The name \"%s\" is not a valid branch name.\n"
+			    "A branch name can not:\n"
+			    /* From the comment for bad_ref_char in refs.c */
+			    "    - Have a path component that begins with \".\"\n"
+			    "    - Have a double dot \"..\"\n"
+			    "    - Have an ASCII control character, \"~\", \"^\", \":\" or SP, anywhere\n"
+			    "    - End with a \"/\"\n"
+			    "    - End with \".lock\"\n"
+			    "    - Contain a \"\\\" (backslash)",
 			    opts.new_branch);
 		if (!get_sha1(buf.buf, rev))
 			die("git checkout: branch %s already exists", opts.new_branch);
-- 
1.7.1.251.g92a7

